src/proxy/pipeline/upstream.rs

```diff
@@ -1,232 +1,295 @@
                     _ = shutdown_rx.changed() => {
-                        debug!(log, "Closing Session");
                         return;
                     }
-                };
+                }
             }
         });
-    }
 
-    /// expiration returns the current expiration Instant value
-    pub fn expiration(&self) -> u64 {
-        self.expiration.load(std::sync::atomic::Ordering::Relaxed)
+        upstream_tx
     }
 
-    /// key returns the key to be used for this session in a SessionMap
-    pub fn key(&self) -> (SocketAddr, SocketAddr) {
-        (self.from, self.dest.address)
-    }
+    async fn process_packet(&self, packet: &[u8]) {
+        slog::trace!(self.log, "Received packet"; "from" => self.endpoint.address,
+            "endpoint_addr" => &self.endpoint.address,
+            "contents" => debug::bytes_to_string(packet));
 
-    /// process_recv_packet processes a packet that is received by this session.
-    async fn process_recv_packet(
-        log: &Logger,
-        metrics: &Metrics,
-        sender: &mut mpsc::Sender<Packet>,
-        expiration: &Arc<AtomicU64>,
-        ttl: Duration,
-        packet_ctx: ReceivedPacketContext<'_>,
-    ) {
-        let ReceivedPacketContext {
-            packet,
-            filter_manager,
-            endpoint,
-            from,
-            to,
-        } = packet_ctx;
-
-        trace!(log, "Received packet"; "from" => from,
-            "endpoint_addr" => &endpoint.address,
-            "contents" => debug::bytes_to_string(&packet));
-
-        if let Err(err) = Session::do_update_expiration(expiration, ttl) {
-            warn!(log, "Error updating session expiration"; "error" => %err)
+        if let Err(err) = self.update_expiration() {
+            slog::warn!(self.log, "Error updating session expiration"; "error" => %err)
         }
 
         let filter_chain = {
-            let filter_manager_guard = filter_manager.read();
+            let filter_manager_guard = self.filter_manager.read();
             filter_manager_guard.get_filter_chain()
         };
-        if let Some(response) =
-            filter_chain.write(WriteContext::new(endpoint, from, to, packet.to_vec()))
-        {
-            if let Err(err) = sender.send(Packet::new(to, response.contents)).await {
-                metrics.rx_errors_total.inc();
-                error!(log, "Error sending packet to channel"; "error" => %err);
+
+        if let Some(response) = filter_chain.write(WriteContext::new(
+            self.endpoint.clone(),
+            self.endpoint.address,
+            self.source_addr,
+            packet.to_vec(),
+        )) {
+            slog::debug!(
+                self.log,
+                "Sending packet back to origin";
+                "origin" => self.endpoint.address,
+                "contents" => debug::bytes_to_string(&response.contents),
+            );
+
+            if let Err(err) = self.downstream_socket.send(&response.contents).await {
```

Comment:
ooh I like this change of sending it directly on the socket rather than the previous queue, good catch!